languages/rust/oso/src/query.rs

```diff
@@ -133,34 +138,39 @@ impl Query {
                 if !kwargs.is_none() {
                     lazy_error!("keyword args for constructor not supported.")
                 } else {
-                    self.host.make_instance(name, args.clone(), instance_id)
+                    let args = args
+                        .iter()
+                        .map(|term| PolarValue::from_term(term, &self.host))
+                        .collect::<crate::Result<Vec<PolarValue>>>()?;
+                    self.host.make_instance(&name.0, args, instance_id)
                 }
             }
             _ => lazy_error!("invalid type for constructing an instance -- internal error"),
         }
     }
 
-    fn register_call(
-        &mut self,
-        call_id: u64,
-        instance: Instance,
-        name: Symbol,
-        args: Option<Vec<Term>>,
-    ) -> crate::Result<()> {
+    fn next_call_result(&mut self, call_id: u64) -> Option<crate::Result<PolarValue>> {
+        self.calls.get_mut(&call_id).and_then(|c| c.next())
+    }
+
+    fn handle_next_external(&mut self, call_id: u64, iterable: Term) -> crate::Result<()> {
         if self.calls.get(&call_id).is_none() {
-            tracing::trace!(call_id, name = %name, args = ?args, "register_call");
-            let results = if let Some(args) = args {
-                instance.call(&name.0, args, &mut self.host)?
-            } else {
-                Box::new(std::iter::once(instance.get_attr(&name.0, &mut self.host)))
-            };
-            self.calls.insert(call_id, results);
+            let instance = Instance::from_polar(PolarValue::from_term(&iterable, &self.host)?)?;
+            let iter = instance
+                .downcast::<crate::host::PolarIterator>(Some(&self.host))
+                .map(|i| Ok(i.clone()))
+                .unwrap_or_else(|_| instance.as_iter(&self.host))?;
+            self.calls.insert(call_id, iter);
```

Comment:
So `self.calls` is now only used for iterators, right? I haven't read the implementation in other language libraries, but should we rename these instance variables so that the name reflects the functionality?

Reply:
Correct. Rename what instance variables?

Reply:
Ohh, like `instance`? Yeah, we can make it `iterable` for example, although that is what `as_iter` is checking.